acvm-repo/brillig_vm/src/lib.rs

```diff
@@ -462,124 +462,211 @@ impl<'a, F: AcirField, B: BlackBoxFunctionSolver<F>> VM<'a, F, B> {
+                                )?;
+                                current_pointer = MemoryAddress(current_pointer.to_usize() + 1);
+                            }
+                            HeapValueType::Array { .. } => {
+                                let destination = self.memory.read_ref(current_pointer);
+                                let destination = self.memory.read_ref(destination);
+                                self.write_slice_of_values_to_memory(
+                                    destination,
+                                    values,
+                                    values_idx,
+                                    typ,
+                                )?;
+                                current_pointer = MemoryAddress(current_pointer.to_usize() + 1);
+                            }
+                            HeapValueType::Vector { .. } => {
+                                return Err(format!(
+                                    "Unsupported returned type in foreign calls {:?}",
+                                    typ
+                                ));
+                            }
+                        }
+                    }
+                }
+                Ok(current_pointer)
+            }
+            HeapValueType::Vector { .. } => {
+                Err(format!("Unsupported returned type in foreign calls {:?}", value_type))
+            }
+        }
+    }
+
+    fn write_value_to_memory(
+        &mut self,
+        destination: MemoryAddress,
+        values: &Vec<F>,
+        values_idx: &mut usize,
+        value_bit_size: u32,
+    ) -> Result<(), String> {
+        if *values_idx >= values.len() {
+            return Err("Missing returned values, you should make sure the foreign call returns a list of flattened values".to_string());
+        }
+        let memory_value = MemoryValue::new_checked(values[*values_idx], value_bit_size);
+
+        if let Some(memory_value) = memory_value {
+            self.memory.write(destination, memory_value);
+        } else {
+            return Err(format!(
+                "Foreign call result value {} does not fit in bit size {}",
+                values[*values_idx], value_bit_size
+            ));
+        }
+        *values_idx += 1;
+        Ok(())
+    }
+
     fn write_foreign_call_result(
         &mut self,
         destinations: &[ValueOrArray],
         destination_value_types: &[HeapValueType],
         foreign_call_index: usize,
     ) -> Result<(), String> {
         let values = &self.foreign_call_results[foreign_call_index].values;
-
         if destinations.len() != values.len() {
             return Err(format!(
                 "{} output values were provided as a foreign call result for {} destination slots",
                 values.len(),
                 destinations.len()
             ));
         }
+        let flatten_values = if values.is_empty() { Vec::new() } else { values[0].fields() };
```

Comment:
Shouldn't this be per-values instead of hardcoded only to be values[0]? that way we can return [...otherStuff, Field[][]] from oracles, and not just [Field[][], ...otherStuff]

Reply:
done